### PR TITLE
Add requirement.txt, modify RolloutWorker

### PR DESCRIPTION
**Pull request by harnvo**
The requirements in README.md is insufficient. 
When downloading dependencies, protobuf with version>=4.0.0 would be installed, which would raise an error. One need to downgrade protobuf to 3.19 after installing pysc2 and smac.
Also, simply run "pip install smac" would not work, for there is  another library called smac. The correct way to do this is to use "pip install git+https://github.com/oxwhirl/smac.git" instead.
In light if this, I freeze my environment after succeed building it. One may use "pip install -r requirements.txt" to install everything correctly.

NOTE: after this, one also need to go to [SMAC](https://github.com/oxwhirl/smac) to put maps into the correct starcraft2 directory.

### CODE REVIEW

**Comment by harnvo** [comment]
Also, there is one slight problem in RolloutWorker. When doing rollout, one need not calculate the gradient of the tensors. In light of this, one need to use @torch.no_grad() wrapper to save memory. See [Pytorch documentation](https://pytorch.org/docs/stable/generated/torch.no_grad.html) for further explanation.

**Comment by starry-sky6688** [comment]
这个requirements.txt文件是用pipreqs生成的么？为什么里面的库是很多是这个项目里没用到的呢，我这边merge的话需要保障requirements.txt的准确性

**Comment by harnvo** [comment]
我是直接pip freeze的。很多库项目里没用到是因为在这些库是pysc2和smac的dependency。不过我直接pip freeze好像确实不太严谨，让我先重新搭建一下环境看看requirements的准确性吧。

**Comment by harnvo** [comment]
> 这个requirements.txt文件是用pipreqs生成的么？为什么里面的库是很多是这个项目里没用到的呢，我这边merge的话需要保障requirements.txt的准确性

我是直接pip freeze的。很多库项目里没用到是因为在这些库是pysc2和smac的dependency。不过我直接pip freeze好像确实不太严谨，让我先重新搭建一下环境看看requirements的准确性吧。

**Comment by harnvo** [comment]
我刚刚看了一下 发现这些库是 pysc2 和 matplotlib 的dependency。这是`pip install pysc2` 后终端上写的： `Successfully installed absl-py-1.2.0 charset-normalizer-2.1.1 deepdiff-5.8.1 dm-env-1.5 dm-tree-0.1.7 enum34-1.1.10 idna-3.3 mock-4.0.3 mpyq-0.2.5 numpy-1.23.2 ordered-set-4.1.0 portpicker-1.5.2 protobuf-4.21.5 psutil-5.9.1 pygame-2.1.2 pysc2-4.0.0 requests-2.28.1 s2clientprotocol-5.0.10.88500.0 s2protocol-5.0.10.88500.0 scipy-1.9.0 sk-video-1.1.10 urllib3-1.26.12 websocket-client-1.3.3`
这是`pip install matplotlib`后终端上写的：`cycler-0.11.0 fonttools-4.37.0 kiwisolver-1.4.4 matplotlib-3.5.3 packaging-21.3 pillow-9.2.0 pyparsing-3.0.9 python-dateutil-2.8.2 six-1.16.0`
我一个个检查过了 和 requirements.txt 里面的全对上了。所以其实requirements 写的应该没有问题。
不过这个freeze的确实太严格了，我重新写了一个对其他库版本没有那么严格的requirements_soft.txt。之前对版本要求严格的变成了requirements.hard.txt。我重新搭建环境试过了，两个都是可以的。

**Comment by starry-sky6688** [comment]
ok，那还是只用soft吧，毕竟大家肯定不想上来先装那么多库，还是越精炼越好；

你把hard删了，然后soft改回成requirements.txt，我来merge

**Comment by starry-sky6688** [comment]
忘说了，soft里还差一个numpy吧？

**Comment by harnvo** [comment]
numpy 是matplotlib 的 dependency

**Comment by harnvo** [comment]
btw 由于smac是从github上下载，直接pip install有可能因为防火墙原因连不上或者找不到smac....这种情况直接重试就行。这个可能需要麻烦您在README里面提一下...太坑了
